### PR TITLE
Primitive's VJP takes outputs as input

### PR DESCRIPTION
**Pull request by awni**
## Proposed changes

This avoids redundant work when a primitive reuses it's output in the VJP.

For example, for the following we now only do one softmax instead of two:

```
def fun(a):
    return mx.sum(mx.softmax(a))

out = mx.grad(fun)(mx.array([1.0, 1.0, 1.0]))
```

### CODE REVIEW

**Comment by awni** [comment]
I don't think I caught all of the cases where we can reuse outputs. If you know of any that I missed, let me know / feel free to push directly.

**Comment by awni** [comment]
Some benchmarks:

## Pre


Linear regression
Loss 0.00005, |w-w*| = 0.00307, Throughput 3703.68598 (it/s)

Logistic regression
Loss 0.00005, |w-w*| = 0.00335, Throughput 3645.06964 (it/s)

Mnist CPU
Epoch 9: Test accuracy 0.945, Time 0.080 (s)

Mnist GPU
Epoch 9: Test accuracy 0.935, Time 0.171 (s)

Transformer LM Batch Size 2
Iter 20: Train loss 8.733, It/sec 2.288

Transformer LM Batch Size 4
Iter 20: Train loss 8.769, It/sec 1.195

Transformer LM Batch Size 8
Iter 20: Train loss 8.873, It/sec 0.558


## Post

Linear regression
Loss 0.00004, |w-w*| = 0.00337, Throughput 3690.08327 (it/s)

Logistic regression
Loss 0.00004, |w-w*| = 0.00334, Throughput 3622.79504 (it/s)

Mnist CPU
Epoch 9: Test accuracy 0.950, Time 0.077 (s)

Mnist GPU
Epoch 9: Test accuracy 0.955, Time 0.169 (s)

Transformer LM Batch Size 2
Iter 20: Train loss 8.929, It/sec 2.294

Transformer LM Batch Size 4
Iter 20: Train loss 8.819, It/sec 1.198

Transformer LM Batch Size 8
Iter 20: Train loss 8.838, It/sec 0.571

**Comment by awni** [comment]
Speech commands small arch

Pre: Throughput: 1678.71 samples/second
Post: Throughput: 1719.39 samples/second

**Review by angeloskath — APPROVED** [review]
Looks fantastic.

I added power, sqrt and scatter. I think these are all.

**Comment by angeloskath** [comment]
Apparently I also broke everything in the process 😅.

Fixing it and pushing again.

**Comment by angeloskath** [comment]
Should be good.
